src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,102 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int ORIGINAL_SIZE = 10;
+    private Object[] elements;
+    private int size;
+
+    public ArrayList() {
+        elements = new Object[ORIGINAL_SIZE];
+    }
+
     @Override
     public void add(T value) {
-
+        if (elements.length == size) {
+            grow();
+        }
+        elements[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        checkIndexForAdd(index);
+        if (elements.length == size) {
+            grow();
+        }
 
+        System.arraycopy(elements, index, elements, index + 1, size - index);
+        elements[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return (T) elements[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        checkIndex(index);
+        elements[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        T value = (T) elements [index];
+        System.arraycopy(elements, index + 1, elements, index, size - index - 1);
+        size--;
+        return value;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        for (int i = 0; i < size; i++) {
+            if (element == elements[i] || element != null && element.equals(elements[i])) {
+                T value = (T) elements[i];
+                return remove(i);
+            }
+        }
+        throw new NoSuchElementException("Elements do not exist" + element);
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private T [] grow() {
+        T[] sourceArray = (T[]) elements;
+        elements = new Object[size + (size >> 1)];
+        System.arraycopy(sourceArray, 0, elements, 0, size);
+        return sourceArray;
+    }
+
+    private void checkIndex(int index) {
+        if (index >= size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index:" + index + "out of size" + size);
+        }
+    }
+
+    private void checkIndexForAdd(int index) {
```

Comment:
you use this method only one time. Put this logic in the place where you call this method